### ISSUE TITLE
On the output layer and use of LSTM in the decoder (Lesson 1)

### ISSUE THREAD

**Issue by cristianoBY** [bentrevett/pytorch-seq2seq#24]
Hi, 

First, Thank you for these clear and illustrative notebooks, and I really appreciate your efforts.

I have two questions regarding the decoder in Lesson 1:

1. The final linear layer of the decoder project the output of the LSTM `output = [sent len, batch size, hid dim * n directions]` to vocab size `prediction = [batch size, output dim]`. However, do we need a softmax layer to normalize the probability over the whole vocab and choose the max as the predicted word? Why can we directly take the max after the linear layer?

2. In the `forward` method of the decoder LSTM, the input is `self.rnn(embedded, (hidden, cell))`, where `#embedded = [1, batch size, emb dim]` is a sequence of length 1. I understand that the forward prop of the decoder must be performed by each time step separately (as seen in the loop in the forward method of the `Seq2Seq` class). My question is that, would there be any difference if we use the `LSTMCell` pytorch class in the decoder? I am pretty sure they will do the same job.

Best,

**Comment by bentrevett** [comment]
Thanks for the kind words.

1. You're correct that `prediction` is un-normalized. The "softmax layer" is part of the `criterion` (loss function), see: https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss

2. Yep, we could use an `nn.LSTMCell` instead of the `nn.LSTM` and it should give the exact same results (if they are both initialized with the exact same values). I wasn't aware of the `LSTMCell` module when first implementing these notebooks, hence why I just used the standard `LSTM` module - I may change it to the cell version at a later date as it avoids all the squeezing/unsqueezing.
